Implement Magnitude-Preserving Orthogonal Ablation

spikymoth opened the pull request:
That is, https://huggingface.co/blog/grimjim/projected-abliteration and https://huggingface.co/blog/grimjim/norm-preserving-biprojected-abliteration with the updated name from https://huggingface.co/posts/grimjim/803126534676334 :)

Marking this as a draft mostly because I don't know the timeline for https://github.com/p-e-w/heretic/pull/43 - either this will need to be rebased on that, or that on this. But the code changes here are relatively small, so I think either direction should be simple enough.

A few notes:
1. The magnitude preservation and orthogonal projection parts can be applied as independent refinements, so I added 2 options for them. From the 2nd article, you need the magnitude preservation for the biprojection to be better than regular projection.
2. The heuristics applied in the article are replaced by heretic's random trials.
3. I did not add magnitude sparsification, which is more of a pre-filtering step.
4. heretic chooses (randomly) one of two strategies for each trial: 1. A refusal direction from a random point between 2 layers, 2. Per-layer refusal directions. The algorithm described in the article is more similar to the former: A refusal direction is chosen from a particular source layer, and then it is projected onto a target layer's harmless direction (this is the biprojection, as far as I can tell).
5. In the article, 2 layers are chosen based on their heuristics. In heretic's random trials, only one layer is chosen - so you'd probably need 2 passes to match the original.

Additional caveats:
1. I'm not a mathematician or have a particularly deep knowledge of matrix math (it doesn't really come up in my day job), so this needs checking to see if I interpreted things correctly.
2. I've only lightly tested this to make sure results don't look broken (refusals go down, KL divergence looks reasonable); I haven't done verification of any metrics. Still running tests locally (another reason for the draft status), but advice on what to check would be welcome.
3. From reading discussion around the method, I know there's some question about the applicability of KL divergence as a metric, since bigger changes may be "good". I've made no attempt to address that here.
4. For a 2nd pass to work, you probably want to ignore parts of the training set that no longer result in refusal. I have some code for that (which seems to work quite well regardless of this change), but I'll make a separate PR for it.

mateusz-malicki commented:
> some layers don't meaningfully separate residuals for good and bad prompts, so they cannot be self-abliterated. But then some layers in between such layers are important for model behavior, and absolutely do need to be abliterated

Hmmm... I was experimenting with modeling ablation weight as polynomials. While it does not solves the main issue, it can help with such cases. huge downside of my approach is the number of parameters we must add - 200 trials isn't even close to "enough" anymore (modeling constraints with motpe sucks).

mateusz-malicki commented:
How? Want to penalize "too low" kl-div? Better idea is to penalize too high refusals count on kl score, so those regions will be less atractive from sampler perspective - but it is still not the best idea im my opinion. Weighting kl-div vs refusals is hard, thats why you get pareto front from motpe instead of best result from tpe.

spikymoth commented:
> How? Want to penalize "too low" kl-div?

I'm not penalizing it per se, just making very low values less interesting to iterate on by pushing the minimum reachable score away from 0. Look at the Pareto front for the last study:
<img width="1014" height="713" alt="image" src="https://github.com/user-attachments/assets/dde92fa9-475a-4152-9366-7d4259b87d8b" />

You can see that it's really enamored with the idea of pushing KL divergence to 0, practically ignoring pushing for a lower number of refusals in comparison. Maybe my fix is too crude and will show the same clustering; checking it now.

mateusz-malicki commented:
@p-e-w Oh, I totally agre on the fact that multi objective is better idea :)

> Weighting kl-div vs refusals is hard, thats why you get pareto front from motpe instead of best result from tpe.

For exactly the same reason as here: We need arbitrary number (offset for kl-div in spikymoths solution or kl-div weight in case of [SO]TPE). And my point is: there is no such universally good number that will fit any model - its model specific; so you have to discover it; so we can see it as additional objective; and we already have motpe in place, so we can skip discovering perfect magic number and just let exploration of pareto front do its job ;) And there is additional point against penalization: there is really no guarantee that our search space is that smooth.

spikymoth commented:
https://github.com/p-e-w/heretic/pull/60 changes things significantly; I'll have to figure out how to apply this in a LoRA context.

spikymoth commented on src/heretic/main.py:
```diff
@@ -380,11 +385,22 @@ def objective(trial: Trial) -> tuple[float, float]:
+                )
+            else:
+                # Because preserving the magnitude reduces the risk from large adjustments,
+                # much larger weight values become viable. However, we still want to
```
Well, I don't think they're mathematically equivalent, at least.

In the original method we subtract the orthogonal portion of the refusal direction, so activations become weaker for weight between [0.8, 1.0] and stronger again for weight between [1.0, 1.5].

With magnitude preservation, we preserve the original activation strengths regardless of the weight, which means the modifications can only affect the alignment with the (negated) refusal direction. Because the orthogonal component of the refusal direction is subtracted along the columns, but the renormalization is done along the rows, it isn't exactly equivalent to a degree of rotation, but it does still reduce the change to a directional one.

Honestly though, I find it pretty hard to reason about. As a modification of the technique, it could also be interesting to preserve the magnitudes along the columns, turning the updates into genuine rotations. Then re-normalize along the rows and restore the activation strengths after that (which would once again change the alignment of the columns, but in a different way).

spikymoth replied:
I tried various configurations, but nothing was better than the original. I did find that normalizing along the columns *first* (then along the rows), and restoring the column norms at the end (after restoring the row norms) also seemed to work okay - but there wasn't a clear advantage to it (if anything it seemed to limit the effectiveness somewhat, but by a surprisingly small amount).

I'm not sure what the best comment would be here, as the adjustments definitely aren't comparable to rotations (normalizing the columns *after* the rows doesn't work at all). I think the initial normalization helps to correlate the columns, and clearly the adjustments are in *some* way constrained by keeping the row norms unchanged, but I'm not sure how to describe it.

p-e-w commented:
The elephant in the room is that once we accept that modifying the model can (or even *should*) modify responses to harmless prompts, there is no reliable way to tell whether things have gotten better or worse. Benchmarks suck, take forever to run, and often contradict human experience with the model.

spikymoth commented on src/heretic/model.py:
```diff
@@ -555,6 +612,13 @@ def get_residuals(self, prompts: list[str]) -> Tensor:
             dim=1,
         )
 
+        if 0 < self.settings.winsorization_level < 100:
```
I suppose there isn't any technical reason why it couldn't be 0, which would set `quantile` to exactly 0.5.

p-e-w replied:
Yes. Also, the level should be expressed as a fraction between 0 and 1. It's unusual to have percentages in code.

spikymoth commented on src/heretic/model.py:
```diff
@@ -442,14 +467,46 @@ def abliterate(
+                        # Note: svd_lowrank actually returns V, not Vh, so transpose it.
+                        U = U[:, :r]
+                        S = S[:r]
+                        Vh = Vh[:, :r].T
```
I was trying to potentially save a bit of memory by reusing the same variable, but I suppose it shouldn't hurt much given the low rank of the LoRA.

spikymoth commented on src/heretic/model.py:
```diff
@@ -145,10 +146,17 @@ def _apply_lora(self):
+            self.lora_rank = 1
+        else:
+            # Magnitude preservation introduces nonlinear effects.
+            self.lora_rank = 3
```
Adding a config setting for this seemed like overkill, especially for trying to get this merged into the main branch (though I'm still experimenting and 3 may not be the optimal value). But if I turn this into a plugin it would make sense to add it (though we'd have to make sure the plugin architecture supports setting it!)

p-e-w replied:
Yeah, this will become a plugin and we definitely want it to be configurable then.

spikymoth commented:
Mostly addressed the comments so far and refined the logic a little more:
* Now uses `torch.linalg.vector_norm`, as `torch.norm` is marked as deprecated
* Clamps the minimum values of the norms to `1e-12` before the division to match `torch.nn.functional.normalize` and avoid potential division by 0
* Adjusts the SVD oversampling to `2 * r + 4`, which should scale a little better for small ranks (more relevant if I make it configurable)

p-e-w commented on src/heretic/model.py:
```diff
@@ -153,10 +155,19 @@ def _apply_lora(self):
+            r=self.lora_rank,
             target_modules=target_modules,
-            lora_alpha=1,
+            lora_alpha=self.lora_rank,
```
This is important. Training code usually sets `lora_alpha` to *twice* the LoRA rank, but in our case we don't want that, because we compute the matrices directly so we need the scaling factor (which is the quotient of those two numbers) to be 1.

p-e-w commented on src/heretic/model.py:
```diff
@@ -410,6 +422,15 @@ def abliterate(
 
+                if self.settings.abliteration_orthogonal_project:
+                    # Remove only the harmful part of the refusal direction.
+                    harmless_direction = harmless_directions[layer_index + 1]
```
This doesn't seem right. If a layer-independent global refusal direction is chosen (`direction_index` is set), we can't use a layer-dependent direction here, that's mixing things up.

spikymoth replied:
This is part of the method as far as I can tell. You pick a specific layer from which to get the refusal direction (like `direction_index`), then for each layer within range, you compute the component of that direction orthogonal to the harmless direction *for that layer*, and subtract that orthogonal component.

I think this is what they mean by "biprojected" - instead of getting the component of the refusal direction orthogonal to the originating layer, you get the component of the refusal direction orthogonal to the *target* layer.

p-e-w replied:
Where is the article describing "biprojected abliteration"? Not the one for "Norm-Preserving Biprojected Abliteration", but the one that describes the original biprojection approach.

https://huggingface.co/blog/grimjim/norm-preserving-biprojected-abliteration says:

> After that, we further refined the technique to "biprojected abliteration", which also removed the corresponding component when removing refusal measured using one layer from another layer entirely

But I can't find an article describing "biprojected abliteration" in any detail.

spikymoth replied:
I was stuck on it for quite a while and tried to get several AIs to tell me what the heck they might have meant with biprojection. But I think what I said is it. [Their code](https://github.com/jim-plus/llm-abliteration/blob/main/sharded_ablate.py#L318) (which I only looked at just now) appears to confirm this:

```python
                    refusal_dir = measures[f'refuse_{measurement}'].float()
                    harmless_dir = measures[f'harmless_{layer}'].float()

                    if projected:
                        # Here we orthogonalize refusal with respect to harmless direction.
                        # We compute the second orthogonalized vector from Gram-Schmitt orthonormalization.

                        # Normalize harmless direction
                        harmless_normalized = torch.nn.functional.normalize(harmless_dir, dim=0)

                        # Project and subtract to refine refusal direction
                        projection_scalar = refusal_dir @ harmless_normalized
                        refined_refusal_dir = refusal_dir - projection_scalar * harmless_normalized
                        refusal_dir = refined_refusal_dir.to(precision)
```

So they use some pre-determined refusal direction, but refine it by subtracting the component aligned with the harmless direction of the layer itself (which I think ultimately amounts to the same thing as what my code does).

p-e-w replied:
Right, but why? Why is the refusal direction global, but the harmlessness direction is per-layer? What is the logic behind this?

Also, where is the "bi" part in "biprojection"? AFAICT, we are only projecting *once*, so why "biprojection"?

p-e-w replied:
I left a comment on the article asking for clarification on this. I'm not comfortable adding logic that I don't understand.

spikymoth replied:
I think the original question was already answered (matching the layers makes more sense and performs better), but regarding global versus per-layer: The method seems to work fine for both, though in my tests the global direction had an advantage (but that might depend on the model).

p-e-w commented on src/heretic/model.py:
```diff
@@ -410,6 +422,15 @@ def abliterate(
+                    layer_refusal_direction -= projection_scalar * harmless_direction
+                    layer_refusal_direction = F.normalize(
+                        layer_refusal_direction, p=2, dim=0
+                    )
```
This logic is independent of the abliteration parameters. If projection is enabled, it should be done *once*, immediately after the residuals are obtained in `main.py`. That will make things not only faster but also simpler, as we can get rid of the `harmless_directions` method parameter. The refusal directions are modified right at the start, and all later code just uses those.

p-e-w commented on src/heretic/model.py:
```diff
@@ -450,14 +478,47 @@ def abliterate(
+                    W = W.view(W.shape[0], -1)
+
+                    if self.settings.abliteration_preserve_magnitude:
+                        # Get a copy of the weight matrix so we can subtract it later.
```
Didn't we do that above already?

p-e-w commented on src/heretic/model.py:
```diff
@@ -450,14 +478,47 @@ def abliterate(
+                        # Apply the LoRA directly.
+                        W.add_(lora_B @ lora_A)
+                        # Normalize the adjusted weight matrix along the rows.
+                        F.normalize(W, p=2, dim=1, out=W)
```
Why does this use `out` instead of assigning the result?

p-e-w commented on src/heretic/model.py:
```diff
@@ -575,6 +636,14 @@ def get_residuals(self, prompts: list[Prompt]) -> Tensor:
 
+        if 0 <= self.settings.winsorization_quantile < 1:
+            # Winsorize the residuals.
+            abs_residuals = torch.abs(residuals).to(torch.float32)
```
The residuals are already being upcast in the return statement below; probably better to move that before this block.

p-e-w commented:
If this PR includes code copied from the original grimjim articles or their GitHub repository (or from anywhere else), please make sure to credit each copied portion, with a link to the original.

spikymoth commented:
> If this PR includes code copied from the original grimjim articles or their GitHub repository (or from anywhere else), please make sure to credit each copied portion, with a link to the original.

I didn't copy any code, but it would make sense to link to the article in a comment.

spikymoth commented on src/heretic/main.py:
```diff
@@ -380,11 +385,22 @@ def objective(trial: Trial) -> tuple[float, float]:
+                max_weight = trial.suggest_float(
+                    f"{component}.max_weight",
+                    0.8,
+                    20.0,
```
I've been playing with this more, and I'm not sure the sampler can make proper use of such large values unless you narrow the parameter range of at least the direction index (which tends to converge fairly obviously). So I might revert this or at least reduce it substantially. I did generally see lower KL divergence with this method, so more aggressive ranges are possible.

p-e-w replied:
I think it's best to remove it from this PR. It's pretty ad-hoc, and hard to understand what precisely is going on regarding the interactions with magnitude normalization.

p-e-w commented on config.default.toml:
```diff
@@ -34,6 +34,12 @@ max_batch_size = 128
+abliteration_orthogonal_project = false
+
+# Whether to preserve the magnitude of the weights.
+abliteration_preserve_magnitude = false
```
I don't like these names very much. They are overly long, prefixed with a scope (`abliteration`) which we don't do elsewhere, and there is also a voice mismatch between `orthogonal_project` (verb last) and `preserve_magnitude` (verb first). Haven't thought of good alternatives yet though.

p-e-w replied:
I avoided that on purpose because namespacing complicates changing settings via the CLI. For prompts it makes sense because there are multiple objects with the same schema, but just to namespace we shouldn't introduce nesting IMO.

(Note that in the future, this will all live inside the namespace of the modifier plugin anyway.)

spikymoth replied:
I changed the names to `orthogonalize_direction`, `orthogonalize_same_layer` (see below) and `preserve_magnitudes`. But I guess it's all temporary really.

anrp commented:
Drive-by anecdote; for reasons unclear and unknown, using heretic to abliterate Qwen3-VL-30B-A3B-Thinking doesn't perform great (minimum refusal count >50 with 100 initial tries and 400 total tries) but using both orthogonal project and preserve magnitude options from this PR greatly decreses refusals down to single digits.

anrp commented:
Some more data, this time openai/gpt-oss-120b; more concrete data, both were 200 initial runs/500 total runs.

[gptoss120b-std.txt](https://github.com/user-attachments/files/24596590/gptoss120b-std.txt)
[gptoss120b-ortho.txt](https://github.com/user-attachments/files/24596592/gptoss120b-ortho.txt)

spikymoth commented on src/heretic/model.py:
```diff
@@ -437,7 +458,14 @@ def abliterate(
-                        W = base_weight.to(torch.float32)
+                        if self.settings.abliteration_preserve_magnitude:
+                            if base_weight.dtype == torch.float32:
+                                # We need a copy for in-place modification.
```
Ah, actually we don't need a copy anymore since there's no more in-place modification. Removed the copy.

spikymoth commented:
Hmm, at first glance (a single run on Gemma-3-27b-it), with
* `orthogonalize_direction = true`
* `orthogonalize_same_layer = false`
* `preserve_magnitudes = true`

(so matching the article), the minimum number of refusals is much higher than before the latest changes.

So the extended range for `max_weight` might be crucial, even if it's hard to explain how it works. But I'll keep testing, also with other parameter combinations.

spikymoth commented:
For easier experimentation, I pushed a commit that adds 2 more settings:
* `max_weight_limit` (default value `1.5`)
* `max_weight_log_scale` (default value `false`)

Happy to remove these again later, but I think they'll come in handy for now.
For reference, the logic of the earlier version was equivalent to:
* `max_weight_limit = 15`
* `max_weight_log_scale = true`

To fully match the previous version and the article (modulo the local/global direction scope):
* `orthogonalize_direction = true`
* `orthogonalize_same_layer = false`
* `preserve_magnitudes = true`
* `max_weight_limit = 15`
* `max_weight_log_scale = true`

spikymoth commented:
The normalization itself is a full rank update, but we just use it to "train" the LoRA; we don't actually store the row-normalized matrix. Also, normalization just means dividing each row by the row norm, and we can undo it with the inverse operation of multiplying each row with the original norm.

The LoRA is still just a rank-1 update, but if we pre-normalize the weight matrix, then we "train" the LoRA to apply to the normalized weight matrix (as `W' = W + ΔW = W + B @ A`). If we want to apply it to the unnormalized matrix, we can just apply the same operation that we would use to restore the norms of the weight matrix: Multiply each row of `ΔW` with the original norm.

But in `ΔW = B @ A`, `B` is a column vector with the same size as the column of row norms, so we can just do an elementwise multiplication `B' = W_norm_rows * B` and then store the "denormalized" `B'`.

I hope that clears it up. In practice, doing this seems to greatly reduce KL divergence while still showing strong refusal reduction (perhaps modestly less strong - but in my experiments with Gemma 3, high divergence results were always incoherent anyway, so low refusal results were just due to invalid refusal measurements).

Doing the full magnitude preservation reduces KL divergences even more by preserving activation strengths, but introduces a tradeoff between keeping the norms unchanged and keeping the original adaptation. The rank-3 truncated SVD minimizes the norm changes for that rank, which most closely aligns with the original intent of re-normalizing without caring what this does to the directional changes. You could do some sort of joint optimization instead, but the SVD approach seems to work well in practice.

p-e-w commented:
I'm trying to wrap my head around how this works geometrically.

AFAICT, the LoRA approach approximates the full operation from norm-preserving biprojected abliteration as a rank-1 (or rank-3) update. This effectively discards higher-order effects from normalization.

You say that

> In practice, doing this seems to greatly reduce KL divergence while still showing strong refusal reduction

But why would we expect this to be the case? This would seem to indicate that the higher-order components of the operation are somehow damaging to the model (in the KLD sense).

Perhaps what's going on is that this "flattens" the operation and makes it more similar to not doing magnitude normalization at all (i.e., to Heretic's original abliteration approach). But note that the point of norm-preserving biprojected abliteration (according to its author) is to make the model smarter even on harmless prompts. In other words, there is an *expected* increase in KLD. I'm wondering if the low-rank approach might be hurting precisely the advantage of this abliteration technique.

spikymoth commented:
Well, to be clear, with `compute_normalized = true` and `preserve_magnitudes = false` we explicitly do *not* preserve row norms. We compute the rank-1 LoRA exactly without any approximation (there are no higher-order components), but we compute it relative to the row-normalized weight matrix.

By computing it this way, the effect of each row on the projection is normalized, which prevents rows with large magnitudes from dominating the resulting projections.

But this is fundamentally different from magnitude preservation, which can theoretically be done even without this pre-normalization step (compute the LoRA relative to the unnormalized weight matrix, then restore the original magnitudes).

I've found that removing the pre-normalization step also makes the magnitude preservation significantly less effective - but I think that's because they're affecting different aspects of the process.

As for expected high KL divergences in the original article - I just don't see those with the optimization process in Heretic. Instead, applying the method seems to *reduce* KL divergence overall, while still effectively reducing refusals.

spikymoth commented:
If it helps, here's the code with pre-normalization if we were to make it unconditional and ignored magnitude preservation:
```python
# Get the row norms (cast to work around untyped LA).
W_norm_row = cast(Tensor, LA.vector_norm(W, dim=1, keepdim=True))
# Normalize the weight matrix along the rows.
W = F.normalize(W, p=2, dim=1)

# Calculate lora_A = v^T W
# v is (d_out,), W is (d_out, d_in)
# v @ W -> (d_in,)
lora_A = (v @ W).view(1, -1)
# Calculate lora_B = -weight * v
# v is (d_out,)
lora_B = (-weight * v).view(-1, 1)

# Apply the original magnitudes to LoRA component B.
lora_B = W_norm_row * lora_B

# LoRA gets applied to unmodified `W`.
```

p-e-w commented:
I think I understand your argument for `compute_normalized` better now:

Normalizing alone prevents high-norm rows from dominating the ablation process, but the original article goes further in trying to prevent ablation from changing the norms at all. That might be going too far, because it's not clear why a modified matrix should have the same norms as the unmodified one – the operator did change, after all, and it's not like it was norm-preserving before so there's nothing inherently special about its original state.

Does that match your reasoning?

p-e-w commented:
/gemini review

spikymoth commented:
Gotcha. Well, would you be willing to have the 3 options that are currently exposed here (ignoring `max_weight_limit` and `max_weight_log_scale` which I'll remove, and `winsorization_quantile` which seems generally useful)? I still think that grouping them their own section in the config file might make sense for them, maybe `[method]` or something (so they would become properties of `settings.method`).

I could also work (in a separate PR) on making the parameter ranges configurable as you suggested above. One concern with this: Should we deal with different module names? A parameter range that makes sense for one module might not make sense for another (especially if you're tightening the ranges based on past results).

spikymoth commented on src/heretic/model.py:
```diff
@@ -153,10 +155,19 @@ def _apply_lora(self):
         ]
 
+        if not self.settings.preserve_magnitudes:
+            # Rank 1 is sufficient for directional ablation
```
Fixed.

p-e-w commented on src/heretic/config.py:
```diff
@@ -19,6 +19,13 @@ class QuantizationMethod(str, Enum):
+class RowNormalization(str, Enum):
+    NONE = "none"
+    PRE = "pre"
+    # POST = "post"  # Theoretically possible, but provides no advantage.
```
What would this do?

p-e-w commented on src/heretic/config.py:
```diff
@@ -159,6 +185,14 @@ class Settings(BaseSettings):
+    winsorization_quantile: float = Field(
+        default=1.0,
+        description=(
+            "The winsorization to apply to the residuals, expressed as the quantile to clamp to (between 0 and 1). "
```
It might be worth pointing out that this process is applied to the individual components of each residual, rather than truncating residual outliers, which could also make sense but isn't what is happening here.

p-e-w commented on src/heretic/main.py:
```diff
@@ -415,11 +415,22 @@ def run():
-    )
+
+    harmless_means = good_residuals.mean(dim=0)
+    harmful_means = bad_residuals.mean(dim=0)
```
I've used the terms "good" and "bad" throughout, rather than "harmful" and "harmless", because they are much easier to distinguish visually in code since they don't both start with the same particle. They also don't carry moral baggage, just a preference of what the user wants. In fact, I'm also going to rename `refusal_directions` eventually to make it more generic.

p-e-w commented on src/heretic/model.py:
```diff
@@ -593,7 +642,17 @@ def get_residuals(self, prompts: list[Prompt]) -> Tensor:
+            thresholds = torch.quantile(
+                abs_residuals, self.settings.winsorization_quantile, 2, True
+            )
+            return torch.clamp(residuals, -thresholds, thresholds)
```
Not sure if I'm reading this correctly. `residuals` has shape `(prompt, layer, component)`. It seems that this code applies separate clamping per prompt, but uniformly across layers. If so, that seems wrong, because residual magnitudes can vary dramatically across layers. What we want (I think) is the reverse: Per-layer clamping, uniformly across all prompts.

p-e-w reviewed (commented):
Full review. Looks excellent overall.

spikymoth commented on src/heretic/model.py:
```diff
@@ -466,6 +477,19 @@ def abliterate(
+                    if self.settings.row_normalization != RowNormalization.NONE:
+                        # Keep a reference to the original weight matrix so we can subtract it later.
+                        W_org = W
+                        # Get the row norms (cast to work around untyped LA).
```
The annoying thing is that annotations don't propagate:
<img width="525" height="484" alt="image" src="https://github.com/user-attachments/assets/9ed031dd-b362-4052-8cbd-2c15ce0d87a3" />
It always implicitly keeps the `Unknown` in there and then that bleeds into subsequent operations. It doesn't actually cause any warnings, though.

spikymoth replied:
I removed the cast in the latest version. Adding the type annotation didn't do anything (as explained above), so I didn't add one.

p-e-w replied:
Which type checker are you using? I think with `ty`, those annotations do propagate as expected.

p-e-w commented on src/heretic/model.py:
```diff
@@ -475,6 +499,31 @@ def abliterate(
+                        # Make the LoRA adapter apply to the original weight matrix.
+                        lora_B = W_row_norms * lora_B
+                    elif self.settings.row_normalization == RowNormalization.FULL:
+                        # Approximates https://huggingface.co/blog/grimjim/norm-preserving-biprojected-abliteration
```
I'm currently developing an entirely new ablation technique that doesn't rely on any assumptions about the structure of the refusal subspace. I hope that approach will not only perform better than the existing ones, but also allow answering more scientifically what the critical rank actually is.

spikymoth commented on src/heretic/model.py:
```diff
@@ -166,10 +168,19 @@ def _apply_lora(self):
+            # Row magnitude preservation introduces nonlinear effects. A rank of 3 is enough to explain
+            # most of the variance in the delta matrix, and reduction of the spectral norm of the error
+            # of the reconstructed matrix falls off at higher ranks.
+            self.lora_rank = 3
```
Setting added.

p-e-w commented:
Merged! This is a milestone for the project, and has been the most-requested feature on Reddit. Thank you for this excellent PR, and for seeing it through to the end.

I plan to ship this as-is in the upcoming Heretic 1.2, and possibly enable it by default in the version after that.
